includes/Services/HashService.php

```diff
@@ -3,55 +3,141 @@
 namespace PowerBoard\Services;
 
 use PowerBoard\PowerBoardPlugin;
+use phpseclib\Crypt\AES;
+use phpseclib\Crypt\Base;
 
 class HashService {
-	private const CIPHER = 'AES-128-CBC';
+
+	private const CIPHER = 'aes-128-cbc';
 	private const OPTION = OPENSSL_RAW_DATA;
+	private const NONCE_LENGTH = 24;
+	private const KEY_LENGTH = 32;
 
 	public static function encrypt( string $string ): string {
 
-		if ( ! function_exists( 'openssl_encrypt' ) ) {
-			return base64_encode( $string );
-		}
+		if ( function_exists( 'sodium_crypto_secretbox' ) ) {
 
-		$ivlen = openssl_cipher_iv_length( self::CIPHER );
-		$iv = openssl_random_pseudo_bytes( $ivlen );
-		$ciphertext_raw = openssl_encrypt( $string, self::CIPHER, self::getKey(), self::OPTION, $iv );
-		$hmac = hash_hmac( 'sha256', $ciphertext_raw, self::getKey(), true );
+			$key = self::getKey( self::KEY_LENGTH );
+			$nonce = random_bytes( self::NONCE_LENGTH );
+			$ciphertext = sodium_crypto_secretbox( $string, $nonce, $key );
 
-		return base64_encode( $iv . $hmac . $ciphertext_raw );
-	}
+			return 'sodium:' . base64_encode( $nonce . $ciphertext );
+
+		} elseif ( class_exists( 'phpseclib\Crypt\AES' ) ) {
+
+			$aes = new AES( AES::MODE_CBC );
+			$aes->setPreferredEngine( Base::ENGINE_INTERNAL );
+			$key = self::getKey( 16 );
+			$ivlen = 16;
+			$iv = random_bytes( $ivlen );
+			$aes->setKey( $key );
+			$aes->setIV( $iv );
+			$ciphertext_raw = $aes->encrypt( $string );
+			$hmac = hash_hmac( 'sha256', $ciphertext_raw, $key, true );
+
+			return 'phpseclib:' . base64_encode( $iv . $hmac . $ciphertext_raw );
+
+		} elseif ( function_exists( 'openssl_encrypt' ) ) {
+
+			$ivlen = openssl_cipher_iv_length( self::CIPHER );
+			$iv = openssl_random_pseudo_bytes( $ivlen );
+			$key = self::getKey( 16 );
+			$ciphertext_raw = openssl_encrypt( $string, self::CIPHER, $key, self::OPTION, $iv );
+			$hmac = hash_hmac( 'sha256', $ciphertext_raw, $key, true );
+
+			return 'openssl:' . base64_encode( $iv . $hmac . $ciphertext_raw );
+
+		} else {
+
+			throw new \Exception( 'There is no available data encryption module' );
```

Comment:
@andryes What happens in the UI if this exception is thrown?

Reply:
@andryes The story states to encrypt merchant ids as well as good practice - i may be wrong but i don't see that, was it not possible?